src/hotspot/share/classfile/moduleEntry.cpp

```diff
@@ -467,27 +484,58 @@ void ModuleEntry::init_as_archived_entry() {
   ArchivePtrMarker::mark_pointer((address*)&_location);
 }
 
-void ModuleEntry::init_archived_oops() {
+void ModuleEntry::update_oops_in_archived_module(int root_oop_index) {
   assert(DumpSharedSpaces, "static dump only");
-  oop module_obj = module();
-  if (module_obj != NULL) {
-    oop m = HeapShared::find_archived_heap_object(module_obj);
-    assert(m != NULL, "sanity");
-    _archived_module_index = HeapShared::append_root(m);
-  }
+  assert(_archived_module_index == -2, "must be set exactly once");
+  assert(root_oop_index >= 0, "sanity");
+
+  _archived_module_index = root_oop_index;
+
   assert(shared_protection_domain() == NULL, "never set during -Xshare:dump");
   // Clear handles and restore at run time. Handles cannot be archived.
   OopHandle null_handle;
   _module = null_handle;
+
+  // For validate_archived_module_entries()
+  DEBUG_ONLY(_num_inited_module_entries++);
+}
+
+char* ModuleEntry::debug_info() const {
+  stringStream info;
+
+  info.print("ModuleEntry* " PTR_FORMAT " (loader = ", p2i(this));
+
+  assert(loader_data()->is_builtin_class_loader_data(), "must be");
+
+  if (loader_data()->is_boot_class_loader_data()) {
+    info.print("boot");
+  } else if (SystemDictionary::is_platform_class_loader(loader_data()->class_loader())) {
+    info.print("platform");
+  } else if (SystemDictionary::is_system_class_loader(loader_data()->class_loader())) {
+    info.print("system");
+  }
```

Comment:
I think this could be simplified to
`info.print(loader_data()->loader_name());`

Reply:
I changed the code to use `ModuleEntry::print()` instead, which internally calls `loader_data()->loader_name_and_id()`.